### PR TITLE
ENV refactor

### PR DESCRIPTION
**Pull request by lasote**
#937 and #939

### CODE REVIEW

File: conans/model/env.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    def sorted_data(self):
+        # Python 3 can't compare None with strings, so if None we order just with the var name
+        return sorted(self._data.items(), key=lambda x: "%s_%s" % x if x[0] else x[1])
+
+    def add(self, name, value, package=None):
+        # It prioritizes the data already introduced
+        if (package, name) not in self._data:
+            self._data[(package, name)] = value
+
+    def global_values(self):
```
Comment (lasote) [inline conans/model/env.py]: Only package_values.

File: conans/client/installer.py [hunk 2]
```diff
@@ -180,6 +180,16 @@ def _propagate_info(self, conan_ref, conan_file, flat):
             conan_file.deps_cpp_info.update(n.conanfile.cpp_info, n.conan_ref)
             conan_file.deps_env_info.update(n.conanfile.env_info, n.conan_ref)
 
+        # Update the env_values with the inherited from dependencies
+        conan_file.env_values.update(conan_file.deps_env_info)
```
Comment (lasote) [inline conans/client/installer.py]: check what

File: conans/client/manager.py [hunk 3]
```diff
@@ -213,8 +217,9 @@ def read_dates(deps_graph):
                     ret[ref] = manifest.time_str
             return ret
 
+        # !! FIXE: Missing env variables, same as install, from profiles etc
```
Comment (lasote) [inline conans/client/manager.py]: yes

File: conans/client/command.py [hunk 4]
```diff
@@ -961,8 +968,8 @@ def run(self, *args):
             logger.error(exc)
             errors = True
         except ConanException as exc:
-            # import traceback
-            # logger.debug(traceback.format_exc())
+            import traceback
+            logger.debug(traceback.format_exc())
```
Comment (lasote) [inline conans/client/command.py]: I know. Just forgot to comment lines again. I usually use them during development.

File: conans/client/deps_builder.py [hunk 5]
```diff
@@ -116,11 +117,13 @@ def propagate_info(self):
...
 
                 non_devs = self.non_dev_nodes(node)
+
                 conanfile.info = ConanInfo.create(conanfile.settings.values,
                                                   conanfile.options.values,
                                                   direct_reqs,
                                                   indirect_reqs,
-                                                  non_devs)
+                                                  non_devs,
+                                                  EnvValues())
```
Comment (lasote) [inline conans/client/deps_builder.py]: Many changes to finally get the current solution, it lost the sense in the way.

File: conans/client/printer.py [hunk 6]
```diff
@@ -176,18 +176,22 @@ def print_search_packages(self, packages_props, reference, recipe_hash, packages
     def print_profile(self, name, profile):
         self._out.info("Configuration for profile %s:\n" % name)
         self._print_profile_section("settings", profile.settings)
-        self._print_profile_section("env", profile.env)
+
+        envs = profile.env_values.global_values()
```
Comment (lasote) [inline conans/client/printer.py]: Not so obvious, the printer uses colored lines and shit like that.

File: conans/model/env.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
+from collections import OrderedDict
+from conans.util.log import logger
+import os
+import re
+
+
+class EnvValues(object):
+    """ Object to represent the introduced env values entered by the user
+    with the -e or profiles etc.
+        self._data is a dictionary with: {(package, var): value}. "package" can be None if the var is global.
```
Comment (lasote) [inline conans/model/env.py]: Ok, I can try it, not sure if there is a good reason, just decided to do that way.

File: conans/client/installer.py (hunk 2 above)
Comment (memsharded) [inline conans/client/installer.py]: Sorry, this was for my-self during review, I wanted to understand.

File: conans/model/env.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        """
+        if env_obj:
+            if isinstance(env_obj, EnvValues):
+                for (package_name, name), value in env_obj._data.items():
+                    self.add(name, value, package_name)
+            else:  # DepsEnvInfo
+                for (name, value) in env_obj.vars.items():
+                    name = name.upper() if name.lower() == "path" else name
+                    if isinstance(value, list):
+                        value = os.pathsep.join(value)
```
Comment (lasote) [inline conans/model/env.py]: Hurray! Comments in code always a good idea, especially for people that are not "the present yourself"

File: conans/model/env.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    def deps(self):
+        return self._dependencies_.keys()
+
+    def __getitem__(self, item):
+        return self._dependencies_[item]
+
+    def update(self, dep_env_info, conan_ref):
+        self._dependencies_[conan_ref.name] = dep_env_info
+
+        def merge_lists(seq1, seq2):
```
Comment (lasote) [inline conans/model/env.py]: Ok, I will push with the **env_info** name, just remember to have the same deference when I review your code :hankey:

**Comment by lasote** [comment]
I've changed the data structure of env_info to {package_name: {env:value}}.
Please review it again.

File: conans/tools.py [hunk 10]
```diff
@@ -23,14 +23,17 @@
...
+    except KeyError:
+        pass
     yield
     sys.path = old_path
 
 
 @contextmanager
-def environment_append(env_vars, keep_vars=["PATH", "PYTHONPATH"]):
+def environment_append(env_vars, keep_vars=None):
+    keep_vars = keep_vars or ["PATH", "PYTHONPATH"]
```
Comment (memsharded) [inline conans/tools.py]: Beware of keep_vers being an empty list. How do you specify you dont want to keep any var?

File: conans/model/profile.py [hunk 11]
```diff
@@ -86,9 +77,9 @@ def get_package_name_value(item):
                         raise ConanException("Invalid env line '%s'" % env)
                     package_name, name, value = get_package_name_value(env)
                     if package_name:
-                        obj._package_env[package_name][name] = value
+                        obj.env_values.add(name, value, package_name)
```
Comment (memsharded) [inline conans/model/profile.py]: You upvoted the comment, but lately didn't move this to EnvValues. Forgot about it or something else I am missing?

Comment (lasote) [inline conans/model/profile.py]: Sorry I forgot that.

**Pull request merged**